File: transparent-bg-skill/scripts/remove-bg.py
#!/usr/bin/env python3
"""
remove-bg.py — 图片去背景透明化工具

支持两种模式：
  1. AI 模式（默认）：使用 rembg 进行 AI 抠图，效果最好
  2. 颜色模式：基于背景颜色替换，适合纯色背景

用法：
  # AI 模式（推荐，首次使用会自动安装 rembg）
  python remove-bg.py input.png
  python remove-bg.py input.png -o output.png

  # 颜色模式（不依赖 AI 模型）
  python remove-bg.py input.png --mode color
  python remove-bg.py input.png --mode color --tolerance 30
  python remove-bg.py input.png --mode color --bg-color "#FFFFFF"

  # 批量处理
  python remove-bg.py *.png --mode ai
  python remove-bg.py img1.png img2.jpg img3.png -o outdir/

  # 边缘羽化（柔化边缘，减少硬边）
  python remove-bg.py input.png --feather 2

参数：
  input           输入图片路径（支持多个）
  -o, --output    输出路径（单文件时为文件名，多文件时为目录）
  --mode          去背景模式：ai（默认） 或 color
  --tolerance     颜色模式容差 0-255（默认 30）
  --bg-color      颜色模式指定背景色（默认自动检测，格式 #RRGGBB）
  --feather       边缘羽化半径像素（默认 0，推荐 1-3）
  --preview       处理后打开图片预览
"""

import argparse
import glob
import os
import subprocess
import sys
from pathlib import Path


def ensure_pillow():
    """确保 Pillow 已安装"""
    try:
        from PIL import Image
        return True
    except ImportError:
        print("📦 正在安装 Pillow...")
        subprocess.check_call([sys.executable, "-m", "pip", "install", "Pillow", "-q"])
        return True


def ensure_rembg():
    """确保 rembg 已安装"""
    try:
        import rembg
        return True
    except ImportError:
        print("📦 正在安装 rembg（首次使用需下载 AI 模型，可能需要几分钟）...")
        try:
            subprocess.check_call(
                [sys.executable, "-m", "pip", "install", "rembg[cpu]", "-q"]
            )
            return True
        except subprocess.CalledProcessError:
            print("⚠️  rembg 安装失败，将回退到颜色模式")
            return False


def detect_bg_color(img, sample_size=10):
    """
    从图片四角采样，检测背景色。
    返回最常见的颜色作为背景色。
    """
    from collections import Counter

    import numpy as np

    pixels = []
    w, h = img.size
    img_array = np.array(img)

    # 从四个角各采样 sample_size x sample_size 个像素
    regions = [
        (0, sample_size, 0, sample_size),                    # 左上
        (0, sample_size, w - sample_size, w),                # 右上
        (h - sample_size, h, 0, sample_size),                # 左下
        (h - sample_size, h, w - sample_size, w),            # 右下
    ]

    for (y1, y2, x1, x2) in regions:
        patch = img_array[y1:y2, x1:x2]
        for row in patch:
            for pixel in row:
                pixels.append(tuple(pixel[:3]))

    # 将颜色量化到 8 的倍数，减少噪声
    quantized = []
    for r, g, b in pixels:
        quantized.append((r // 8 * 8, g // 8 * 8, b // 8 * 8))

    counter = Counter(quantized)
    bg_color = counter.most_common(1)[0][0]
    return bg_color


def color_distance(c1, c2):
    """计算两个 RGB 颜色之间的欧氏距离"""
    return ((c1[0] - c2[0]) ** 2 + (c1[1] - c2[1]) ** 2 + (c1[2] - c2[2]) ** 2) ** 0.5


def remove_bg_color(img, tolerance=30, bg_color=None):
    """
    颜色模式去背景：
    1. 检测/使用指定的背景色
    2. 将与背景色相近的像素设为透明
    3. 使用 flood-fill 从边缘开始，只移除连通的背景区域
    """
    from PIL import Image
    import numpy as np

    img = img.convert("RGBA")
    data = np.array(img)

    if bg_color is None:
        bg_color = detect_bg_color(img)
        print(f"  🎨 检测到背景色: RGB({bg_color[0]}, {bg_color[1]}, {bg_color[2]})")

    # 计算每个像素与背景色的距离
    rgb = data[:, :, :3].astype(float)
    bg = np.array(bg_color, dtype=float)
    dist = np.sqrt(np.sum((rgb - bg) ** 2, axis=2))

    # 距离在容差范围内的标记为背景候选
    bg_mask = dist <= tolerance

    # Flood fill 从四条边开始，只移除与边缘连通的背景
    # 使用 scipy.ndimage.label 加速连通区域检测，回退到纯 numpy 实现
    h, w = bg_mask.shape

    try:
        from scipy.ndimage import label as ndimage_label
        # 标记所有连通的背景区域
        labeled, num_features = ndimage_label(bg_mask)
        # 收集与边缘接触的区域标签
        edge_labels = set()
        edge_labels.update(labeled[0, :].flatten())      # 上边
        edge_labels.update(labeled[h - 1, :].flatten())  # 下边
        edge_labels.update(labeled[:, 0].flatten())      # 左边
        edge_labels.update(labeled[:, w - 1].flatten())  # 右边
        edge_labels.discard(0)  # 0 表示非背景
        # 只保留与边缘连通的背景
        visited = np.isin(labeled, list(edge_labels))
    except ImportError:
        # 纯 numpy BFS 回退
        from collections import deque
        visited = np.zeros_like(bg_mask, dtype=bool)
        queue = deque()

        # 将四条边上的背景候选像素加入队列
        for x in range(w):
            if bg_mask[0, x] and not visited[0, x]:
                queue.append((0, x))
                visited[0, x] = True
            if bg_mask[h - 1, x] and not visited[h - 1, x]:
                queue.append((h - 1, x))
                visited[h - 1, x] = True
        for y in range(h):
            if bg_mask[y, 0] and not visited[y, 0]:
                queue.append((y, 0))
                visited[y, 0] = True
            if bg_mask[y, w - 1] and not visited[y, w - 1]:
                queue.append((y, w - 1))
                visited[y, w - 1] = True

        # BFS flood fill with deque for performance
        while queue:
            cy, cx = queue.popleft()
            for dy, dx in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                ny, nx = cy + dy, cx + dx
                if 0 <= ny < h and 0 <= nx < w and not visited[ny, nx] and bg_mask[ny, nx]:
                    visited[ny, nx] = True
                    queue.append((ny, nx))

    # 将连通背景区域设为透明
    data[visited, 3] = 0

    # 对边缘进行半透明过渡（anti-alias）
    try:
        from scipy.ndimage import binary_dilation
        edge = binary_dilation(visited, iterations=1) & ~visited
        edge_ys, edge_xs = np.where(edge)
        for y, x in zip(edge_ys, edge_xs):
            d = dist[y, x]
            if d < tolerance:
                alpha_ratio = d / tolerance
                data[y, x, 3] = int(255 * alpha_ratio)
    except ImportError:
        pass  # scipy 不可用时跳过边缘优化

    return Image.fromarray(data)


def remove_bg_ai(img):
    """AI 模式去背景：使用 rembg"""
    from rembg import remove
    return remove(img)


def apply_feather(img, radius):
    """对 alpha 通道进行羽化处理"""
    if radius <= 0:
        return img
    from PIL import ImageFilter
    import numpy as np

    img = img.convert("RGBA")
    data = np.array(img)
    alpha = data[:, :, 3]

    # 找到边缘区域并模糊 alpha
    from PIL import Image as PILImage

    alpha_img = PILImage.fromarray(alpha)
    # 对 alpha 通道做高斯模糊
    blurred = alpha_img.filter(ImageFilter.GaussianBlur(radius=radius))
    blurred_data = np.array(blurred)

    # 只对边缘区域应用模糊（既不是完全透明也不是完全不透明的区域扩展）
    # 简化：直接用模糊后的 alpha，但保持完全不透明的内部区域
    mask_opaque = alpha == 255
    mask_transparent = alpha == 0

    # 在不透明区域保留原始 alpha，在透明区域保留 0，边缘使用模糊值
    result_alpha = np.where(mask_opaque, 255, np.where(mask_transparent, blurred_data, blurred_data))
    # 进一步：完全透明区域如果模糊后仍然很低就保持 0
    result_alpha = np.where(mask_transparent & (blurred_data < 10), 0, result_alpha)

    data[:, :, 3] = result_alpha.astype(np.uint8)
    from PIL import Image as PILImage2
    return PILImage2.fromarray(data)


def process_single(input_path, output_path, mode, tolerance, bg_color, feather, preview):
    """处理单张图片"""
    from PIL import Image

    input_path = Path(input_path)
    if not input_path.exists():
        print(f"❌ 文件不存在: {input_path}")
        return False

    print(f"🖼️  处理: {input_path.name}")

    img = Image.open(input_path).convert("RGBA")
    print(f"  📐 尺寸: {img.size[0]}x{img.size[1]}")

    if mode == "ai":
        result = remove_bg_ai(img)
        print("  ✨ AI 去背景完成")
    else:
        result = remove_bg_color(img, tolerance=tolerance, bg_color=bg_color)
        print(f"  ✨ 颜色去背景完成 (容差={tolerance})")

    if feather > 0:
        result = apply_feather(result, feather)
        print(f"  🌫️  边缘羽化: {feather}px")

    # 确保输出目录存在
    output_path = Path(output_path)
    output_path.parent.mkdir(parents=True, exist_ok=True)

    # 保存为 PNG（支持透明度）
    if output_path.suffix.lower() not in [".png", ".webp"]:
        output_path = output_path.with_suffix(".png")

    result.save(str(output_path), "PNG")
    file_size = output_path.stat().st_size
    print(f"  💾 保存: {output_path} ({file_size / 1024:.1f} KB)")

    if preview:
        try:
            if sys.platform == "win32":
                os.startfile(str(output_path))
            elif sys.platform == "darwin":
                subprocess.run(["open", str(output_path)])
            else:
                subprocess.run(["xdg-open", str(output_path)])
        except Exception:
            pass

    return True


def generate_output_path(input_path, output_arg, is_batch):
    """生成输出路径"""
    input_path = Path(input_path)

    if output_arg:
        output = Path(output_arg)
        if is_batch or output.is_dir() or str(output_arg).endswith(("/", "\\")):
            # 输出到目录
            output.mkdir(parents=True, exist_ok=True)
            stem = input_path.stem
            if not stem.endswith("_transparent"):
                stem += "_transparent"
            return output / f"{stem}.png"
        else:
            return output
    else:
        # 默认：同目录，文件名加 _transparent 后缀
        stem = input_path.stem
        if stem.endswith("_transparent"):
            return input_path.parent / f"{stem}.png"
        return input_path.parent / f"{stem}_transparent.png"


def parse_color(color_str):
    """解析颜色字符串 #RRGGBB"""
    if not color_str:
        return None
    color_str = color_str.strip("#")
    if len(color_str) != 6:
        raise ValueError(f"Invalid color format: #{color_str}, expected #RRGGBB")
    r = int(color_str[0:2], 16)
    g = int(color_str[2:4], 16)
    b = int(color_str[4:6], 16)
    return (r, g, b)


def main():
    parser = argparse.ArgumentParser(
        description="图片去背景透明化工具",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
示例:
  %(prog)s photo.png                         # AI 模式去背景
  %(prog)s photo.png --mode color            # 颜色模式去背景
  %(prog)s *.png -o transparent/             # 批量处理到指定目录
  %(prog)s icon.png --feather 2 --preview    # 羽化边缘 + 预览
        """,
    )

    parser.add_argument("input", nargs="+", help="输入图片路径（支持通配符）")
    parser.add_argument("-o", "--output", help="输出路径（文件或目录）")
    parser.add_argument(
        "--mode",
        choices=["ai", "color"],
        default="ai",
        help="去背景模式：ai（默认，AI 抠图）或 color（颜色替换）",
    )
    parser.add_argument(
        "--tolerance",
        type=int,
        default=30,
        help="颜色模式容差 0-255（默认 30，越大去除越多）",
    )
    parser.add_argument(
        "--bg-color",
        help="颜色模式指定背景色（格式 #RRGGBB，默认自动检测）",
    )
    parser.add_argument(
        "--feather",
        type=int,
        default=0,
        help="边缘羽化半径（默认 0，推荐 1-3）",
    )
    parser.add_argument(
        "--preview", action="store_true", help="处理后打开图片预览"
    )

    args = parser.parse_args()

    # 展开通配符
    input_files = []
    for pattern in args.input:
        expanded = glob.glob(pattern)
        if expanded:
            input_files.extend(expanded)
        else:
            input_files.append(pattern)

    if not input_files:
        print("❌ 未找到任何输入文件")
        sys.exit(1)

    is_batch = len(input_files) > 1
    print(f"🔧 模式: {'AI 抠图' if args.mode == 'ai' else '颜色替换'}")
    print(f"📁 共 {len(input_files)} 个文件待处理\n")

    # 确保依赖
    ensure_pillow()
    if args.mode == "ai":
        if not ensure_rembg():
            print("⚠️  回退到颜色模式")
            args.mode = "color"

    bg_color = parse_color(args.bg_color) if args.bg_color else None

    success = 0
    failed = 0

    for input_file in input_files:
        output_path = generate_output_path(input_file, args.output, is_batch)
        ok = process_single(
            input_file,
            output_path,
            mode=args.mode,
            tolerance=args.tolerance,
            bg_color=bg_color,
            feather=args.feather,
            preview=args.preview,
        )
        if ok:
            success += 1
        else:
            failed += 1
        print()

    print(f"{'=' * 40}")
    print(f"✅ 成功: {success}  ❌ 失败: {failed}  📊 总计: {len(input_files)}")


if __name__ == "__main__":
    main()

File: transparent-bg-skill/scripts/remove-bg.mjs
#!/usr/bin/env node
/**
 * remove-bg.mjs — 图片去背景透明化工具（Node.js 版）
 *
 * 支持两种模式：
 *   1. AI 模式（默认）：使用 @imgly/background-removal-node AI 抠图，效果最好
 *   2. 颜色模式：基于背景色 flood-fill 替换，适合纯色背景，无需大型依赖
 *
 * 依赖（全局安装，不污染项目）：
 *   npm install -g @imgly/background-removal-node   # AI 模式
 *   npm install -g sharp                             # 颜色模式 + 图片 I/O
 *
 * 用法：
 *   node remove-bg.mjs input.png
 *   node remove-bg.mjs input.png -o output.png
 *   node remove-bg.mjs input.png --mode color --tolerance 30
 *   node remove-bg.mjs "素材库/download/*.png" -o "素材库/download/transparent/"
 *   node remove-bg.mjs sprite.png --feather 2 --preview
 */

import { readFileSync, writeFileSync, existsSync, mkdirSync, statSync } from 'fs';
import { createRequire } from 'module';
import { join, dirname, basename, extname, resolve } from 'path';
import { fileURLToPath } from 'url';
import { globSync } from 'fs';

const __dirname = dirname(fileURLToPath(import.meta.url));

// ─── CLI 参数解析 ───────────────────────────────────────────────
function parseArgs() {
  const args = process.argv.slice(2);
  const opts = {
    inputs: [],
    output: null,
    mode: 'ai',
    tolerance: 30,
    bgColor: null,
    feather: 0,
    preview: false,
  };

  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === '-o' || a === '--output') { opts.output = args[++i]; }
    else if (a === '--mode') { opts.mode = args[++i]; }
    else if (a === '--tolerance') { opts.tolerance = parseInt(args[++i], 10); }
    else if (a === '--bg-color') { opts.bgColor = args[++i]; }
    else if (a === '--feather') { opts.feather = parseInt(args[++i], 10); }
    else if (a === '--preview') { opts.preview = true; }
    else if (a === '--help' || a === '-h') { printHelp(); process.exit(0); }
    else if (!a.startsWith('-')) { opts.inputs.push(a); }
  }
  return opts;
}

function printHelp() {
  console.log(`
remove-bg.mjs — 图片去背景透明化工具

用法：
  node remove-bg.mjs <input...> [options]

选项：
  -o, --output PATH     输出路径（单文件=文件名，多文件=目录）
  --mode {ai,color}     去背景模式（默认 ai）
  --tolerance N         颜色模式容差 0-255（默认 30）
  --bg-color #RRGGBB    颜色模式指定背景色（默认自动检测）
  --feather N           边缘羽化半径像素（默认 0，推荐 1-3）
  --preview             处理后打开图片预览
  --help                显示帮助

示例：
  node remove-bg.mjs photo.png
  node remove-bg.mjs photo.png --mode color --tolerance 25
  node remove-bg.mjs "*.png" -o transparent/
  node remove-bg.mjs icon.png --feather 2 --preview
`);
}

// ─── 依赖加载 ───────────────────────────────────────────────────
const requireGlobal = createRequire(
  join(process.env.APPDATA || process.env.HOME, 'npm', 'node_modules', '.package-lock.json')
);

function tryLoadSharp() {
  // 1. 本地
  try { return (await import('sharp')).default; } catch {}
  // 只能用 require 方式全局
  return null;
}

async function loadSharp() {
  // 先尝试本地
  try {
    const { default: s } = await import('sharp');
    return s;
  } catch {}
  // 全局路径候选
  const candidates = [
    join(process.env.APPDATA || '', 'npm', 'node_modules', 'sharp'),
    join(process.env.HOME || '', '.npm', 'lib', 'node_modules', 'sharp'),
    '/usr/local/lib/node_modules/sharp',
  ];
  for (const p of candidates) {
    if (existsSync(p)) {
      try {
        const s = createRequire(p + '/package.json')('sharp');
        return s;
      } catch {}
    }
  }
  return null;
}

async function loadRemoveBg() {
  // 先尝试本地
  try {
    const m = await import('@imgly/background-removal-node');
    return m.removeBackground ?? m.default?.removeBackground;
  } catch {}
  // 全局路径候选
  const candidates = [
    join(process.env.APPDATA || '', 'npm', 'node_modules', '@imgly', 'background-removal-node'),
    join(process.env.HOME || '', '.npm', 'lib', 'node_modules', '@imgly', 'background-removal-node'),
    '/usr/local/lib/node_modules/@imgly/background-removal-node',
  ];
  for (const p of candidates) {
    if (existsSync(p)) {
      try {
        const pkg = join(p, 'dist', 'node.cjs');
        if (existsSync(pkg)) {
          const r = createRequire(pkg);
          const m = r(pkg);
          return m.removeBackground ?? m.default?.removeBackground;
        }
      } catch {}
    }
  }
  return null;
}

// ─── 颜色工具 ───────────────────────────────────────────────────
function parseColor(hex) {
  if (!hex) return null;
  const h = hex.replace('#', '');
  if (h.length !== 6) throw new Error(`颜色格式错误: ${hex}，需为 #RRGGBB`);
  return [parseInt(h.slice(0, 2), 16), parseInt(h.slice(2, 4), 16), parseInt(h.slice(4, 6), 16)];
}

function colorDist(a, b) {
  return Math.sqrt((a[0]-b[0])**2 + (a[1]-b[1])**2 + (a[2]-b[2])**2);
}

// ─── 背景色检测（从四角采样） ───────────────────────────────────
function detectBgColor(data, width, height, sampleSize = 10) {
  const pixels = [];
  const regions = [
    [0, sampleSize, 0, sampleSize],
    [0, sampleSize, width - sampleSize, width],
    [height - sampleSize, height, 0, sampleSize],
    [height - sampleSize, height, width - sampleSize, width],
  ];
  for (const [y1, y2, x1, x2] of regions) {
    for (let y = y1; y < Math.min(y2, height); y++) {
      for (let x = x1; x < Math.min(x2, width); x++) {
        const i = (y * width + x) * 4;
        // 量化到 8 的倍数减少噪声
        pixels.push([(data[i] >> 3) << 3, (data[i+1] >> 3) << 3, (data[i+2] >> 3) << 3]);
      }
    }
  }
  const counts = new Map();
  for (const p of pixels) {
    const key = `${p[0]},${p[1]},${p[2]}`;
    counts.set(key, (counts.get(key) || 0) + 1);
  }
  const best = [...counts.entries()].sort((a, b) => b[1] - a[1])[0][0];
  return best.split(',').map(Number);
}

// ─── 颜色模式：flood-fill 从边缘去除背景 ───────────────────────
function removeBgColor(data, width, height, tolerance = 30, bgColor = null) {
  const buf = Buffer.from(data); // 拷贝，不修改原始数据

  const bg = bgColor || detectBgColor(buf, width, height);
  console.log(`  🎨 背景色: RGB(${bg[0]}, ${bg[1]}, ${bg[2]})`);

  const visited = new Uint8Array(width * height);
  // 预计算每像素与背景色距离
  const dist = new Float32Array(width * height);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const i4 = (y * width + x) * 4;
      dist[y * width + x] = colorDist([buf[i4], buf[i4+1], buf[i4+2]], bg);
    }
  }

  // BFS 从所有四条边开始
  const queue = [];
  for (let x = 0; x < width; x++) {
    if (dist[x] <= tolerance && !visited[x]) { visited[x] = 1; queue.push(x); }
    const bot = (height - 1) * width + x;
    if (dist[bot] <= tolerance && !visited[bot]) { visited[bot] = 1; queue.push(bot); }
  }
  for (let y = 1; y < height - 1; y++) {
    const l = y * width;
    if (dist[l] <= tolerance && !visited[l]) { visited[l] = 1; queue.push(l); }
    const r = y * width + width - 1;
    if (dist[r] <= tolerance && !visited[r]) { visited[r] = 1; queue.push(r); }
  }

  let head = 0;
  const dx = [1, -1, 0, 0];
  const dy = [0, 0, 1, -1];
  while (head < queue.length) {
    const idx = queue[head++];
    const cx = idx % width, cy = Math.floor(idx / width);
    for (let d = 0; d < 4; d++) {
      const nx = cx + dx[d], ny = cy + dy[d];
      if (nx < 0 || nx >= width || ny < 0 || ny >= height) continue;
      const ni = ny * width + nx;
      if (visited[ni] || dist[ni] > tolerance) continue;
      visited[ni] = 1;
      queue.push(ni);
    }
  }

  // 将连通背景区域设为透明
  for (let i = 0; i < width * height; i++) {
    if (visited[i]) buf[i * 4 + 3] = 0;
  }

  return buf;
}

// ─── 羽化：对 alpha 通道做简单盒式模糊 ──────────────────────────
function applyFeather(data, width, height, radius) {
  if (radius <= 0) return data;
  const buf = Buffer.from(data);
  const alpha = new Float32Array(width * height);
  for (let i = 0; i < width * height; i++) alpha[i] = buf[i * 4 + 3];

  // 水平方向模糊
  const tmp = new Float32Array(width * height);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      let sum = 0, cnt = 0;
      for (let k = -radius; k <= radius; k++) {
        const nx = x + k;
        if (nx >= 0 && nx < width) { sum += alpha[y * width + nx]; cnt++; }
      }
      tmp[y * width + x] = sum / cnt;
    }
  }
  // 垂直方向模糊
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      let sum = 0, cnt = 0;
      for (let k = -radius; k <= radius; k++) {
        const ny = y + k;
        if (ny >= 0 && ny < height) { sum += tmp[ny * width + x]; cnt++; }
      }
      // 只对边缘区域（非完全透明/非完全不透明）应用模糊
      const orig = alpha[y * width + x];
      const blurred = sum / cnt;
      const result = orig === 255 ? 255 : orig === 0 ? 0 : blurred;
      buf[(y * width + x) * 4 + 3] = Math.round(result);
    }
  }
  return buf;
}

// ─── 输出路径推导 ───────────────────────────────────────────────
function getOutputPath(inputPath, outputArg, isBatch) {
  const ext = extname(inputPath);
  const stem = basename(inputPath, ext);
  if (outputArg) {
    // 判断是否为目录
    if (isBatch || outputArg.endsWith('/') || outputArg.endsWith('\\') ||
        (existsSync(outputArg) && statSync(outputArg).isDirectory())) {
      mkdirSync(outputArg, { recursive: true });
      const s = stem.endsWith('_transparent') ? stem : stem + '_transparent';
      return join(outputArg, s + '.png');
    }
    return outputArg;
  }
  const s = stem.endsWith('_transparent') ? stem : stem + '_transparent';
  return join(dirname(inputPath), s + '.png');
}

// ─── 展开 glob ──────────────────────────────────────────────────
function expandInputs(inputs) {
  const files = [];
  for (const p of inputs) {
    if (p.includes('*') || p.includes('?')) {
      try {
        const matches = globSync(p);
        if (matches.length) { files.push(...matches); continue; }
      } catch {}
    }
    files.push(p);
  }
  return files;
}

// ─── 预览 ───────────────────────────────────────────────────────
function openPreview(filePath) {
  const { execSync } = await import('child_process');
  const abs = resolve(filePath);
  try {
    if (process.platform === 'win32') execSync(`start "" "${abs}"`);
    else if (process.platform === 'darwin') execSync(`open "${abs}"`);
    else execSync(`xdg-open "${abs}"`);
  } catch {}
}

// ─── 处理单张图片 ───────────────────────────────────────────────
async function processSingle(inputPath, outputPath, opts, sharp, removeBg) {
  if (!existsSync(inputPath)) {
    console.log(`❌ 文件不存在: ${inputPath}`);
    return false;
  }

  console.log(`🖼️  处理: ${basename(inputPath)}`);

  // 确保输出目录存在
  mkdirSync(dirname(resolve(outputPath)), { recursive: true });

  if (opts.mode === 'ai') {
    if (!removeBg) {
      console.log('❌ AI 模式需要全局安装 @imgly/background-removal-node');
      console.log('   npm install -g @imgly/background-removal-node');
      return false;
    }
    const inputBuf = readFileSync(inputPath);
    const blob = new Blob([inputBuf], { type: 'image/png' });
    const result = await removeBg(blob, { publicPath: 'https://staticimgly.com/@imgly/background-removal-data/1.5.3/dist/' });
    const outBuf = Buffer.from(await result.arrayBuffer());

    if (opts.feather > 0) {
      // 羽化需要经过 sharp 解码/编码
      if (!sharp) {
        writeFileSync(outputPath, outBuf);
        console.log('  ⚠️  未安装 sharp，跳过羽化');
      } else {
        const { data, info } = await sharp(outBuf).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
        const feathered = applyFeather(data, info.width, info.height, opts.feather);
        await sharp(feathered, { raw: { width: info.width, height: info.height, channels: 4 } }).png().toFile(outputPath);
        console.log(`  🌫️  边缘羽化: ${opts.feather}px`);
      }
    } else {
      writeFileSync(outputPath, outBuf);
    }
    console.log('  ✨ AI 去背景完成');

  } else {
    // 颜色模式
    if (!sharp) {
      console.log('❌ 颜色模式需要全局安装 sharp');
      console.log('   npm install -g sharp');
      return false;
    }
    const { data, info } = await sharp(inputPath).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
    console.log(`  📐 尺寸: ${info.width}x${info.height}`);

    const bgColor = opts.bgColor ? parseColor(opts.bgColor) : null;
    let result = removeBgColor(data, info.width, info.height, opts.tolerance, bgColor);
    console.log(`  ✨ 颜色去背景完成 (容差=${opts.tolerance})`);

    if (opts.feather > 0) {
      result = applyFeather(result, info.width, info.height, opts.feather);
      console.log(`  🌫️  边缘羽化: ${opts.feather}px`);
    }

    await sharp(result, { raw: { width: info.width, height: info.height, channels: 4 } }).png().toFile(outputPath);
  }

  const size = statSync(outputPath).size;
  console.log(`  💾 保存: ${outputPath} (${(size / 1024).toFixed(1)} KB)`);

  if (opts.preview) await openPreview(outputPath);
  return true;
}

// ─── 主流程 ─────────────────────────────────────────────────────
async function main() {
  const opts = parseArgs();

  if (opts.inputs.length === 0) {
    printHelp();
    process.exit(1);
  }

  const inputFiles = expandInputs(opts.inputs);
  if (inputFiles.length === 0) {
    console.log('❌ 未找到任何输入文件');
    process.exit(1);
  }

  console.log(`🔧 模式: ${opts.mode === 'ai' ? 'AI 抠图' : '颜色替换'}`);
  console.log(`📁 共 ${inputFiles.length} 个文件待处理\n`);

  // 加载依赖
  const sharp = await loadSharp();
  const removeBg = opts.mode === 'ai' ? await loadRemoveBg() : null;

  if (opts.mode === 'ai' && !removeBg) {
    console.log('⚠️  未找到 @imgly/background-removal-node，请先全局安装：');
    console.log('   npm install -g @imgly/background-removal-node');
    console.log('   安装后重新运行，或改用 --mode color\n');
    process.exit(1);
  }
  if (opts.mode === 'color' && !sharp) {
    console.log('⚠️  未找到 sharp，颜色模式需要全局安装：');
    console.log('   npm install -g sharp');
    process.exit(1);
  }

  const isBatch = inputFiles.length > 1;
  let success = 0, failed = 0;

  for (const inputPath of inputFiles) {
    const outputPath = getOutputPath(inputPath, opts.output, isBatch);
    const ok = await processSingle(inputPath, outputPath, opts, sharp, removeBg);
    ok ? success++ : failed++;
    console.log();
  }

  console.log('='.repeat(40));
  console.log(`✅ 成功: ${success}  ❌ 失败: ${failed}  📊 总计: ${inputFiles.length}`);
}

main().catch(e => { console.error('❌', e.message); process.exit(1); });
